File: apps/ios/ExpCore/Sources/API/BoardsApi.swift
import Foundation

/// v4: a board IS a repository. Create/retarget always carry a repo — either
/// an existing team-registry repo (`repositoryId`) or a brand-new repo
/// connected inline by `fullName` (server validates the GitHub-App install and
/// upserts the registry row in the same transaction). Mirrors the server
/// `repositoryInputSchema` (apps/web/src/lib/trpc/boards.ts).
public enum BoardRepositoryChoice: Encodable, Sendable, Equatable {
    /// Target an existing registry repo (same-team, not archived).
    case repositoryId(String)
    /// Connect a new repo inline by `owner/name`; the extra fields seed the
    /// registry row (all optional server-side). The installation id is
    /// resolved server-side from GitHub — clients never send it.
    case fullName(String, defaultBranch: String? = nil, isPrivate: Bool? = nil)

    private enum IdKeys: String, CodingKey { case repositoryId }
    private enum NameKeys: String, CodingKey { case fullName, defaultBranch, `private` }

    public func encode(to encoder: Encoder) throws {
        switch self {
        case let .repositoryId(id):
            var c = encoder.container(keyedBy: IdKeys.self)
            try c.encode(id, forKey: .repositoryId)
        case let .fullName(name, defaultBranch, isPrivate):
            var c = encoder.container(keyedBy: NameKeys.self)
            try c.encode(name, forKey: .fullName)
            try c.encodeIfPresent(defaultBranch, forKey: .defaultBranch)
            try c.encodeIfPresent(isPrivate, forKey: .private)
        }
    }
}

public struct CreateBoardInput: Encodable, Sendable {
    public let teamId: String
    public let name: String
    public let prefix: String
    public var color: String?
    // The curated glyph name.
    public let icon: String?
    // Optional on EVERY board now (the type collapse): coding/PR affordances
    // gate on repo presence, never on a required repo. The server no longer
    // rejects a repo-less board.
    public let repository: BoardRepositoryChoice?
    // EXP-712: the board's own branch — coding sessions start from it and its
    // PRs target it. Only meaningful alongside `repository` (the server drops
    // it otherwise); nil = follow the repo's default branch.
    public let defaultBranch: String?

    public init(
        teamId: String,
        name: String,
        prefix: String,
        color: String? = nil,
        icon: String? = nil,
        repository: BoardRepositoryChoice? = nil,
        defaultBranch: String? = nil
    ) {
        self.teamId = teamId
        self.name = name
        self.prefix = prefix
        self.color = color
        self.icon = icon
        self.repository = repository
        self.defaultBranch = defaultBranch
    }

    enum CodingKeys: String, CodingKey {
        case teamId, name, prefix, color, icon, repository, defaultBranch
    }

    public func encode(to encoder: Encoder) throws {
        var c = encoder.container(keyedBy: CodingKeys.self)
        try c.encode(teamId, forKey: .teamId)
        try c.encode(name, forKey: .name)
        try c.encode(prefix, forKey: .prefix)
        try c.encodeIfPresent(color, forKey: .color)
        try c.encodeIfPresent(icon, forKey: .icon)
        // Omit `repository` entirely when nil so the server's optional schema
        // sees an absent key (not JSON null, which the union would reject).
        try c.encodeIfPresent(repository, forKey: .repository)
        // Same reasoning: absent, never null — the server's `nullish()` would
        // take a null, but omitting keeps create's payload minimal.
        try c.encodeIfPresent(defaultBranch, forKey: .defaultBranch)
    }
}

public struct SetBoardRepositoryInput: Encodable, Sendable {
    public let boardId: String
    /// Null clears the board's repo (coding affordances gate on presence).
    public let repositoryId: String?
    /// EXP-712: a retarget RESETS the board's branch unless one is passed —
    /// a branch belongs to the repo it was picked in.
    public let defaultBranch: String?

    public init(boardId: String, repositoryId: String?, defaultBranch: String? = nil) {
        self.boardId = boardId
        self.repositoryId = repositoryId
        self.defaultBranch = defaultBranch
    }

    enum CodingKeys: String, CodingKey {
        case boardId, repositoryId, defaultBranch
    }

    public func encode(to encoder: Encoder) throws {
        var c = encoder.container(keyedBy: CodingKeys.self)
        try c.encode(boardId, forKey: .boardId)
        // Explicit null: the server reads it as "no repository", so this key
        // can never be omitted the way an optional input would be.
        try c.encode(repositoryId, forKey: .repositoryId)
        try c.encodeIfPresent(defaultBranch, forKey: .defaultBranch)
    }
}

/// `boards.update` — a partial patch. Only the keys a caller sets are sent,
/// except `defaultBranch`, whose explicit null MEANS "follow the repo again"
/// (EXP-712) and so is always written when the caller opted into it.
public struct UpdateBoardInput: Encodable, Sendable {
    public let boardId: String
    public let name: String?
    public let color: String?
    public let icon: String?
    private let branchPatch: Bool
    private let defaultBranch: String?

    public init(
        boardId: String,
        name: String? = nil,
        color: String? = nil,
        icon: String? = nil
    ) {
        self.boardId = boardId
        self.name = name
        self.color = color
        self.icon = icon
        branchPatch = false
        defaultBranch = nil
    }

    /// The branch patch: `defaultBranch: nil` sends JSON null (follow the repo).
    public init(boardId: String, defaultBranch: String?) {
        self.boardId = boardId
        name = nil
        color = nil
        icon = nil
        branchPatch = true
        self.defaultBranch = defaultBranch
    }

    enum CodingKeys: String, CodingKey {
        case boardId, name, color, icon, defaultBranch
    }

    public func encode(to encoder: Encoder) throws {
        var c = encoder.container(keyedBy: CodingKeys.self)
        try c.encode(boardId, forKey: .boardId)
        try c.encodeIfPresent(name, forKey: .name)
        try c.encodeIfPresent(color, forKey: .color)
        try c.encodeIfPresent(icon, forKey: .icon)
        if branchPatch { try c.encode(defaultBranch, forKey: .defaultBranch) }
    }
}

public struct BoardResult: Decodable, Sendable {
    public let board: BoardResultData

    public init(board: BoardResultData) {
        self.board = board
    }
}

public struct BoardResultData: Decodable, Sendable {
    public let id: String

    public init(id: String) {
        self.id = id
    }
}

public final class BoardsApi: Sendable {
    private let trpc: TrpcClient

    public init(trpc: TrpcClient) {
        self.trpc = trpc
    }

    /// Create a board. The server uppercases `prefix` and defaults `color`
    /// to `#6366f1` when omitted. Returns the new board id.
    public func create(accountId: String, _ input: CreateBoardInput) async throws -> String {
        let result: BoardResult = try await trpc.mutation(accountId: accountId, path: "boards.create", input: input)
        return result.board.id
    }

    /// Retarget a board's backing repo (owner/admin, server-enforced). The
    /// repo must already be in the team registry — connecting a brand-new
    /// repo happens via `create`'s inline path or `repositories.add`. Electric
    /// surfaces the updated `repository_id`.
    public func setRepository(
        accountId: String,
        boardId: String,
        repositoryId: String?,
        defaultBranch: String? = nil
    ) async throws {
        try await trpc.mutationVoid(
            accountId: accountId,
            path: "boards.setRepository",
            input: SetBoardRepositoryInput(
                boardId: boardId,
                repositoryId: repositoryId,
                defaultBranch: defaultBranch
            )
        )
    }

    /// Patch a board (`boards.update`). EXP-712: `defaultBranch` nil means
    /// "follow the repo's default branch again" and is sent as JSON null.
    public func setDefaultBranch(accountId: String, boardId: String, defaultBranch: String?) async throws {
        try await trpc.mutationVoid(
            accountId: accountId,
            path: "boards.update",
            input: UpdateBoardInput(boardId: boardId, defaultBranch: defaultBranch)
        )
    }
}

File: apps/android/app/src/main/java/com/exponential/app/data/api/BoardsApi.kt
package com.exponential.app.data.api

import com.exponential.app.data.db.BoardEntity
import javax.inject.Inject
import javax.inject.Singleton
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonNull
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.contentOrNull
import kotlinx.serialization.json.put

// Mirrors apps/web/src/lib/trpc/boards.ts `create`. A repository is OPTIONAL on
// every board (EXP-121); when supplied it's either an existing team-registry
// repo (`repositoryId`) or a brand-new repo connected inline by `fullName` (the
// server validates the GitHub-App install and upserts the registry row in the
// same transaction). Modeled after iOS `BoardRepositoryChoice`.

/** The optional backing-repo choice for `boards.create`. */
sealed interface BoardRepositoryChoice {
    /** Target an existing registry repo (same-team, not archived). */
    data class Registry(val repositoryId: String) : BoardRepositoryChoice

    /**
     * Connect a new repo inline by `owner/name`; extra fields seed the registry
     * row. The installation id is resolved server-side — never sent by clients.
     */
    data class Inline(
        val fullName: String,
        val defaultBranch: String? = null,
        val isPrivate: Boolean? = null,
    ) : BoardRepositoryChoice

    fun toJson(): JsonObject = when (this) {
        is Registry -> buildJsonObject { put("repositoryId", repositoryId) }
        is Inline -> buildJsonObject {
            put("fullName", fullName)
            defaultBranch?.let { put("defaultBranch", it) }
            isPrivate?.let { put("private", it) }
        }
    }
}

@Serializable
private data class CreateBoardResult(val board: JsonObject)

/**
 * `boards.create` input, hand-built so the `repository` union encodes exactly
 * as the server's `z.union` expects (registry vs inline shapes differ) and so
 * an absent optional stays ABSENT rather than riding as a null.
 * `defaultBranch` (EXP-712) is only meaningful with a repository — the server
 * drops it otherwise.
 */
internal fun createBoardInput(
    teamId: String,
    name: String,
    prefix: String,
    color: String?,
    icon: String,
    repository: BoardRepositoryChoice?,
    defaultBranch: String?,
): JsonObject = buildJsonObject {
    put("teamId", teamId)
    put("name", name)
    put("prefix", prefix)
    color?.let { put("color", it) }
    put("icon", icon)
    repository?.let { put("repository", it.toJson()) }
    defaultBranch?.takeIf { it.isNotBlank() }?.let { put("defaultBranch", it) }
}

/**
 * `boards.update`'s branch write (EXP-712). The router applies a field only
 * when the key is PRESENT, so "follow the repo again" has to travel as a
 * literal `null` — and the shared Json's `explicitNulls = false` would drop a
 * null property of a `@Serializable` class, turning the clear into a silent
 * no-op (the ActionsWireFormatTest story).
 */
internal fun updateBoardBranchInput(
    boardId: String,
    defaultBranch: String?,
): JsonObject = buildJsonObject {
    put("boardId", boardId)
    put("defaultBranch", defaultBranch?.let(::JsonPrimitive) ?: JsonNull)
}

/**
 * `boards.create` result: the new board's id (always present) plus the
 * full row when the server response decodes into a [BoardEntity] (the server
 * returns every column via `.returning()`). The entity drives the optimistic
 * local Room upsert (EXP-46); it's null on older/self-hosted servers whose
 * response is missing required fields — callers then just wait for Electric.
 */
data class CreatedBoardResult(val id: String, val entity: BoardEntity?)

@Singleton
class BoardsApi @Inject constructor(
    private val trpc: TrpcClient,
    private val json: Json,
) {

    /**
     * Create a board. The server uppercases `prefix` and defaults `color` to
     * `#6366f1` when omitted. Since the board-type collapse (EXP-121) we send
     * `icon` (curated contract name) instead of the legacy `type`; a
     * `repository` is OPTIONAL on every board. The inline-connect path needs
     * owner/admin (repo management). `defaultBranch` (EXP-712) pins the
     * board's own branch; null follows the repo's default. Returns the new
     * board id plus the full row when decodable (see [CreatedBoardResult]).
     */
    suspend fun create(
        accountId: String,
        teamId: String,
        name: String,
        prefix: String,
        color: String?,
        icon: String,
        repository: BoardRepositoryChoice?,
        defaultBranch: String? = null,
    ): CreatedBoardResult {
        val input: JsonElement =
            createBoardInput(teamId, name, prefix, color, icon, repository, defaultBranch)
        val board = trpc.mutation(
            accountId,
            path = "boards.create",
            input = input,
            inputSerializer = JsonElement.serializer(),
            outputSerializer = CreateBoardResult.serializer(),
        ).board
        val id = (board["id"] as? JsonPrimitive)?.contentOrNull
            ?: throw TrpcException("boards.create returned no board id")
        // Tolerant full-row decode (BoardEntity accepts the tRPC camelCase
        // names via @JsonNames): a server returning fewer fields degrades to
        // id-only instead of failing the already-committed create.
        val entity = runCatching {
            json.decodeFromJsonElement(BoardEntity.serializer(), board)
        }.getOrNull()
        return CreatedBoardResult(id = id, entity = entity)
    }

    /**
     * `boards.update` — the board's own branch (EXP-712). `null` clears the
     * pin, so the board follows its repo's default branch again. Member-level,
     * like every other board field write; the updated row arrives over
     * Electric.
     */
    suspend fun setDefaultBranch(accountId: String, boardId: String, defaultBranch: String?) =
        trpc.mutationUnit(
            accountId,
            path = "boards.update",
            input = updateBoardBranchInput(boardId, defaultBranch),
            inputSerializer = JsonElement.serializer(),
        )
}
